architecture/2023-12-asyncworker-progressInfo.md

```diff
@@ -0,0 +1,130 @@
+# Title
+
+* **Status**: Pending
+* **Author**: Vinaya Damle (@vinayada1)
+
+## Overview
+
+Currently async worker has no knowledge of the operation progress till the operation is complete. If the operation times out, there is no mechanism in the current design that any information about the operation progress so far can be communicated to the user. As a result, the user only sees the error message that the operation timed out.
+
+The proposal here is to add a mechanism so that the running operation can send out progress information to the async worker. As a result, when the deployment operation eventually times out, along with the timed out message, the user will also be able to see the operation progress so far in the error message which could help in troubleshooting the deployment failure.
+
+## Terms and definitions
+
+AsyncWorker - The current architecture which assigns workers to process queued long-running jobs asynchronously.
+
+## Objectives
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6835
+
+### Goals
+- Enable reporting of operation progress in async worker
+
+### Non-Goals
+- Modify reporting of operation progress in CLI
+
+### User scenarios (optional)
+
+#### User story 1
+
+The readiness probe for the application pod fails. This is a non-terminal failure condition. The current code keeps waiting for the deployment to complete and eventually the deployment times out. The error message to the user is "Deployment timed out after xxx seconds".
+
+If we implement https://github.com/radius-project/radius/issues/6284 and look at pod events, we can detect the readiness probe failure. However, we still need a way to report this information to the user.
```

Comment:
Do you have an example of what this would look like? eg: if we deployed a container what set of progress messages would we create?

Would this become the default behavior for all of our async operations? eg: when an operation times out we'd build an error message from the progress updates?